app/src/main/java/com/ivanpodlesnykh/playlistmaker/SearchHistory.kt

```diff
@@ -0,0 +1,66 @@
+package com.ivanpodlesnykh.playlistmaker
+
+import android.content.SharedPreferences
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+class SearchHistory(val sharedPreferences: SharedPreferences) {
+
+    private var trackList: ArrayList<Track> = arrayListOf()
+
+    init {
+        val json = sharedPreferences.getString("SEARCH_HISTORY", null)
+        if(json != null) {
+            val itemType = object : TypeToken<ArrayList<Track>>() {}.type
+            trackList = Gson().fromJson<ArrayList<Track>>(json, itemType)
+        }
+    }
+
+    fun saveTrackToList(track: Track) {
+        var hasDuplicate = false
+
+        var json = sharedPreferences.getString("SEARCH_HISTORY", null)
+        if (json != null) {
+            val itemType = object : TypeToken<ArrayList<Track>>() {}.type
+            trackList = Gson().fromJson<ArrayList<Track>>(json, itemType)
+        }
+
+        for (item in trackList) {
+            if(item.trackId == track.trackId) {
+                trackList.remove(item)
+                hasDuplicate = true
+                break
+            }
+        }
+        if(hasDuplicate) trackList.add(track)
+        else {
+            if(trackList.size < 10) trackList.add(track)
+            else {
+                trackList.removeAt(0)
+                trackList.add(track)
+            }
+        }
+
+        json = Gson().toJson(trackList)
+
+        sharedPreferences.edit().putString("SEARCH_HISTORY", json).apply()
+    }
+
+    fun getTrackList(): ArrayList<Track> {
+        val json = sharedPreferences.getString("SEARCH_HISTORY", null)
+        if (json != null) {
+            val itemType = object : TypeToken<ArrayList<Track>>() {}.type
+            trackList = Gson().fromJson<ArrayList<Track>>(json, itemType)
+            return trackList
+        }
+        return arrayListOf()
```

Comment:
```suggestion
        val json = sharedPreferences.getString("SEARCH_HISTORY", null) ?: return arrayListOf()
        val itemType = object : TypeToken<ArrayList<Track>>() {}.type
        trackList = Gson().fromJson<ArrayList<Track>>(json, itemType)
        return trackList
```
Не будет лишнего ветвления и так будет удобнее читать код